File: collect_app/src/main/java/org/odk/collect/android/injection/config/ProjectDependencyProviders.java
package org.odk.collect.android.injection.config;

import androidx.annotation.NonNull;

import org.odk.collect.android.entities.ProjectJsonFileEntitiesRepository;
import org.odk.collect.android.projects.ProjectDependencyProvider;
import org.odk.collect.android.storage.ProjectStoragePaths;
import org.odk.collect.entities.EntitiesRepository;

import dagger.assisted.AssistedFactory;

public interface ProjectDependencyProviders {
    @AssistedFactory
    interface EntitiesRepositoryProvider extends ProjectDependencyProvider<EntitiesRepository> {
        ProjectJsonFileEntitiesRepository get(@NonNull String projectId);
    }

    @AssistedFactory
    interface ProjectStoragePathsFactory extends ProjectDependencyProvider<ProjectStoragePaths> {
        ProjectStoragePaths get(String projectId);
    }
}

File: collect_app/src/main/java/org/odk/collect/android/injection/config/ProjectDependencyProviders.kt
package org.odk.collect.android.injection.config

import dagger.assisted.AssistedFactory
import org.odk.collect.android.entities.JsonFileEntitiesRepository
import org.odk.collect.android.projects.ProjectDependencyProvider
import org.odk.collect.android.storage.ProjectStoragePaths
import org.odk.collect.entities.EntitiesRepository
import java.io.File
import javax.inject.Inject

object ProjectDependencyProviders {
    class EntitiesRepositoryProvider @Inject constructor(
        private val projectStoragePathsFactory: ProjectStoragePathsFactory
    ) : ProjectDependencyProvider<EntitiesRepository> {
        override fun get(projectId: String): EntitiesRepository {
            val paths = projectStoragePathsFactory.get(projectId)
            return JsonFileEntitiesRepository(File(paths.rootDir))
        }
    }

    @AssistedFactory
    interface ProjectStoragePathsFactory : ProjectDependencyProvider<ProjectStoragePaths> {
        override fun get(projectId: String): ProjectStoragePaths
    }
}
